From 0d141bf93a5c2cda3c95822926e1999e50b0cbfd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?S=C3=B6ren=20Tempel?= <tempel@uni-bremen.de>
Date: Mon, 8 Feb 2021 13:42:18 +0100
Subject: [PATCH] clif: Don't access any data if input is empty

This is relevant as clif_decode_link may invoke clif_get_attr with
input_len == 0.
---
 sys/clif/clif.c                         |  4 ++++
 tests/unittests/tests-clif/tests-clif.c | 11 +++++++++++
 2 files changed, 15 insertions(+)

diff --git a/sys/clif/clif.c b/sys/clif/clif.c
index 893a5cdd9c12..7376a196d99a 100644
--- a/sys/clif/clif.c
+++ b/sys/clif/clif.c
@@ -258,6 +258,10 @@ ssize_t clif_get_attr(const char *input, size_t input_len, clif_attr_t *attr)
     attr->value = NULL;
     attr->key = NULL;
 
+    if (input_len == 0) {
+        return CLIF_NOT_FOUND;
+    }
+
     /* an attribute should start with the separator */
     if (*pos != LF_ATTR_SEPARATOR_C) {
         DEBUG("Attribute should start with separator, found %c\n", *pos);
diff --git a/tests/unittests/tests-clif/tests-clif.c b/tests/unittests/tests-clif/tests-clif.c
index c299bac0ca09..d915cbb7df50 100644
--- a/tests/unittests/tests-clif/tests-clif.c
+++ b/tests/unittests/tests-clif/tests-clif.c
@@ -285,12 +285,23 @@ static void test_clif_get_attr_missing_value(void)
     TEST_ASSERT_EQUAL_INT(strlen(input), r);
 }
 
+static void test_clif_get_attr_empty(void)
+{
+    clif_attr_t attr;
+
+    /* clif_get_attr used to access data even if input was empty.
+     * See: https://github.com/RIOT-OS/RIOT/pull/15947 */
+    int r = clif_get_attr(NULL, 0, &attr);
+    TEST_ASSERT_EQUAL_INT(CLIF_NOT_FOUND, r);
+}
+
 Test *tests_clif_tests(void)
 {
     EMB_UNIT_TESTFIXTURES(fixtures) {
         new_TestFixture(test_clif_encode_links),
         new_TestFixture(test_clif_decode_links),
         new_TestFixture(test_clif_get_attr_missing_value),
+        new_TestFixture(test_clif_get_attr_empty)
     };
 
     EMB_UNIT_TESTCALLER(clif_tests, NULL, NULL, fixtures);
